### PR TITLE
[Bookmark] handling of clicking on link when url starts with #.

### PR DESCRIPTION
**Pull request by pszczesniak**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal: Extended handling of clicking on link when `URL` starts with `#`.

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 1]
```diff
@@ -70,16 +76,20 @@ export default class LinkActionsView extends View {
...
+	private readonly _bookmarkEditing: BookmarkEditing | null;
+
 	declare public t: LocaleTranslate;
 
 	/**
 	 * @inheritDoc
 	 */
-	constructor( locale: Locale, linkConfig: LinkConfig = {} ) {
-		super( locale );
+	constructor( editor: Editor, linkConfig: LinkConfig = {} ) {
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:86]: This introduces a breaking change. I'd suggest not passing the whole editor instance to the view.

**Review by niegowski — CHANGES_REQUESTED** [review]

File: packages/ckeditor5-link/src/ui/linkactionsview.ts (hunk 1 above)
Comment (pszczesniak) [reply packages/ckeditor5-link/src/ui/linkactionsview.ts]: Fair point, i've extended the passed arguments list.

File: packages/ckeditor5-link/lang/contexts.json [hunk 2]
```diff
@@ -8,6 +8,7 @@
 	"Open link in new tab": "Button opening the link in new browser tab.",
 	"This link has no URL": "Label explaining that a link has no URL set (the URL is empty).",
 	"Open in a new tab": "The label of the switch button that controls whether the edited link will open in a new tab.",
+	"Scroll to bookmark": "The label of the switch button that controls whether the edited link will scroll to a bookmark.",
```
Comment (niegowski) [inline packages/ckeditor5-link/lang/contexts.json:11]: I'm not sure if this is correct. Maybe `Scroll to the bookmark`?

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 3]
```diff
@@ -70,16 +71,19 @@ export default class LinkActionsView extends View {
 
 	private readonly _linkConfig: LinkConfig;
 
+	private readonly _bookmarkEditing: BookmarkEditing | null;
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:74]: Maybe a callback would be a more generic approach.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 4]
```diff
@@ -70,16 +71,19 @@ export default class LinkActionsView extends View {
...
+
 	declare public t: LocaleTranslate;
 
 	/**
 	 * @inheritDoc
 	 */
-	constructor( locale: Locale, linkConfig: LinkConfig = {} ) {
+	constructor( locale: Locale, linkConfig: LinkConfig = {}, bookmarkEditing: BookmarkEditing | null = null ) {
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:81]: If we decide to pass a callback, I'd pass it in an `options` object.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 5]
```diff
@@ -220,12 +248,18 @@ export default class LinkActionsView extends View {
...
-		button.template!.eventListeners = {};
 
 		return button;
 	}
 }
 
+/**
+ * Checks if clicking the link can be scrolled to bookmark when bookmark `id` matches hash `url`.
+ */
+function isScrollableToBookmark( link: string | undefined, bookmarkEditing: BookmarkEditing | null ): boolean {
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:259]: Maybe this helper should be provided from the outside world as an options callback.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 6]
```diff
@@ -208,9 +211,34 @@ export default class LinkActionsView extends View {
...
+				rel: bind.to( 'href', href => {
+					if ( isScrollableToBookmark( href, this._bookmarkEditing ) ) {
+						return 'noopener';
+					}
+
+					return 'noopener noreferrer';
+				} )
+			},
+			on: {
+				click: bind.to( () => {
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:230]: Maybe a `evt.preventDefault()` call would be good here so we could ignore other link related attributes for in editable scroll.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 7]
```diff
@@ -208,9 +211,34 @@ export default class LinkActionsView extends View {
...
+					handleLinkOpening( this.href!, this._bookmarkEditing );
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:231]: Maybe we need another callback passed in options so we do not hardcode the bookmark logic in this view class.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 8]
```diff
@@ -208,9 +211,34 @@ export default class LinkActionsView extends View {
...
+				} )
+			}
+		} );
+
+		button.bind( 'tooltip' ).to( this, 'href', href => {
+			if ( isScrollableToBookmark( href, this._bookmarkEditing ) ) {
+				return t( 'Scroll to bookmark' );
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:238]: Maybe `Scroll to target`?

File: packages/ckeditor5-link/src/ui/linkactionsview.ts (hunk 7 above)
Comment (niegowski) [reply packages/ckeditor5-link/src/ui/linkactionsview.ts:231]: Maybe if we describe it as an internal vs external link?

File: packages/ckeditor5-link/src/utils.ts [hunk 9]
```diff
@@ -194,6 +196,35 @@ export function openLink( link: string ): void {
...
+	}
+
+	bookmarkEditing.editor.model.change( writer => {
+		writer.setSelection( modelBookmark, 'before' );
+	} );
+
+	bookmarkEditing.editor.editing.view.scrollToTheSelection( {
+		alignToTop: true,
+		forceScroll: true
+	} );
```
Comment (niegowski) [inline packages/ckeditor5-link/src/utils.ts:225]: This looks like a separate helper, the rest of the code of this function is a duplicate of `isScrollableToBookmark` helper.

File: packages/ckeditor5-link/src/linkediting.ts [hunk 10]
```diff
@@ -280,7 +280,9 @@ export default class LinkEditing extends Plugin {
 			evt.stop();
 			data.preventDefault();
 
-			openLink( url );
+			const bookmarkEditing = editor.plugins.has( 'BookmarkEditing' ) ? editor.plugins.get( 'BookmarkEditing' ) : null;
+
+			handleLinkOpening( url, bookmarkEditing );
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkediting.ts:285]: I'd extract it above those listeners and use sort of a factory to create link opener.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 11]
```diff
@@ -210,17 +212,38 @@ export default class LinkActionsView extends View {
...
 			return href || t( 'This link has no URL' );
 		} );
 
 		button.bind( 'isEnabled' ).to( this, 'href', href => !!href );
 
 		button.template!.tag = 'a';
-		button.template!.eventListeners = {};
+
+		// When `eventListeners` is "cleaned" the binding to the `click` event is removed.
+		// button.template!.eventListeners = {};
```
Comment (pszczesniak) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:246]: When it's uncommented the on click event is not working.

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 12]
```diff
@@ -70,16 +70,19 @@ export default class LinkActionsView extends View {
 
 	private readonly _linkConfig: LinkConfig;
 
+	private readonly _options: Record<string, Function> | undefined;
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:73]:
```suggestion
	private readonly _options?: LinkActionsViewOptions;
```

File: packages/ckeditor5-link/src/ui/linkactionsview.ts [hunk 13]
```diff
@@ -70,16 +70,19 @@ export default class LinkActionsView extends View {
...
+
 	declare public t: LocaleTranslate;
 
 	/**
 	 * @inheritDoc
 	 */
-	constructor( locale: Locale, linkConfig: LinkConfig = {} ) {
+	constructor( locale: Locale, linkConfig: LinkConfig = {}, options?: Record<string, Function> ) {
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkactionsview.ts:80]:
```suggestion
	constructor( locale: Locale, linkConfig: LinkConfig = {}, options?: LinkActionsViewOptions ) {
```

**Review by niegowski — APPROVED** [review]

**Pull request merged**